### ISSUE TITLE
Cursor stay at the top of email

### ISSUE THREAD

**Issue by rbgarga** [robertklep/quotefixformac#51]
When I try to reply an email, it sends me to the bottom, as if I used scroll bar to do it, but cursor stays at the top, then if I press something to start to write reply message, it moves me back to top, where cursor is, and start the type there.

I'm using El Capitan and Quotefix 2.7.0

**Comment by rbgarga** [comment]
Looks like the problem is here: https://github.com/robertklep/quotefixformac/commit/abb99a0126a03587ee339e2427c71ca3ca067fd6?w=0#diff-15110fb3dee6ab7e58fd27b3d352b529R380

I replaced this elif by a simple else statement and it worked

**Comment by houmark** [comment]
+1 — I also noticed that my first typing is getting screwed by the cursor moving around.

Often when I hit reply I start typing right away while the animation is ongoing, and that used to work with no issues, but recently it started messing up the cursor position.

**Comment by robertklep** [comment]
@rbgarga do you mean like this?

```
if self.app.move_cursor_to_top:
    view.moveToBeginningOfDocument_(self)
else:
    view.moveToEndOfDocument_(self)
```

@houmark I'm not sure if I can fix that. QF has to wait for the animation to finish before moving the cursor to the correct position, otherwise the cursor will be positioned at the top of the document. By _"used to work with no issues"_, do you mean on El Capitan, or on Yosemite?

**Comment by houmark** [comment]
@robertklep That's a good question. As far as I remember it was working fine on Yosemite. On El Cap I have noticed the cursor having both this problem and also that it was adding an empty line above the cursor (even without any typing while the animation is ongoing), but that seems better now.

Would it maybe be possible to detect where the cursor is already when starting, and if there's no text (after trimming) or less than say 20-30 chars before the cursor (to accommodate "a bit of typing" while the animation ran), then it should not move? Just thinking out of the box here for potential hacks :)

**Comment by rbgarga** [comment]
@robertklep exactly, maybe a problem with my signature? I can make tests if you need

**Comment by robertklep** [comment]
@rbgarga it sounds like QF may not be able to find the signature properly. Are you using plain Mail signatures, or through a plugin?

@houmark do I understand correctly that you position the cursor yourself (and start typing) before the animation has finished?

**Comment by rbgarga** [comment]
@robertklep plain text signature like this:

--
Renato Botelho

After I changed that statement I mentioned before it's moving cursor to the end of the email, below my signature. It's not a fix, just a way to have cursor near the place I want it :)

**Comment by houmark** [comment]
@robertklep No, I just start typing and then the cursor is moved after a bit of time (like after ~1 second) and that screws up the current word(s) until I notice, correct and move the cursor.

It's hard to explain, but I can reproduce every time if I hit reply and start typing right away. I can maybe try making a short video for you if you would like that?

**Comment by robertklep** [comment]
@houmark I think I understand what's happening. However, if you start typing before the cursor is placed properly, I wonder where you want your reply text to end up in the first place: above or below the quoted text.

Default behaviour for Mail is to place the cursor at the top (above the quoted text), perhaps I should add an option to not mess with the cursor position at all. Only if you want the cursor to be placed below the quoted text do you have to wait for the animation to be completed (and only on ElCap).

@rbgarga if you're okay with changing the code, could you log the outcome of `self.move_above_new_signature(htmldom, view)`? It should return false if it was unable to find the signature. Something like this:

```
NSLog('result: %@', self.move_above_new_signature(htmldom, view))
```

**Comment by rbgarga** [comment]
@robertklep Sure. I would be glad to help. I've added the NSLog call. Where does it save the log?

**Comment by robertklep** [comment]
@rbgarga `/var/log/system.log` (accessible either through Terminal or with the Console app)

**Comment by rbgarga** [comment]
@robertklep result: 1

**Comment by robertklep** [comment]
@rbgarga in that case, it was able to find your signature. I can't really explain the behaviour you're witnessing, especially that the document is scrolled down but the cursor stays on top. I'll have to see if I can somehow reproduce the problem somehow.

**Comment by rbgarga** [comment]
@robertklep Don't ask me why, but I restored original code and it's working as expected now, that is really odd

**Comment by rbgarga** [comment]
I'll let you know if I can reproduce it and send you detailed steps. Thank you for help and congrats for the great plugin!

**Comment by houmark** [comment]
@robertklep Yes, an option like that would be awesome. I want the text to be in the top of the e-mail.

Honestly, I'm using quotefix very simple. I just want it to change the attribution on the reply and nothing else, so if it would be possible to have it not mess with the cursor then that would be a nice option I guess. I'd be happy to test and provide more feedback.

**Comment by robertklep** [comment]
@houmark just to make sure: have you tried enabling the _"Move cursor to top of document after QuoteFixing"_ option?

<img width="695" alt="option" src="https://cloud.githubusercontent.com/assets/1190916/10709897/2446f1ee-7a3f-11e5-930f-65a5b8561aa3.png">

**Comment by houmark** [comment]
@robertklep Sorry for my very late reply on this. I thought I had replied somewhere.

Yes, I have that setting on normally and I also tried without it checked. The result is a bit different for me, but same annoyance. It moves the cursor no matter and screws up the actual writing. I think in some way QuoteFix is adjusting the cursor position no matter if it's on or off, just in different ways, but you probably know that better.

It would be nice with a setting to disable any cursor movement, but I'm afraid it will then maybe show the cursor further down in the mail and then I as a user still would have to move the content upwards or do small tweaks.

After months with this I keep hitting the cursor movement constantly, because I use the keyboard shortcuts to start a reply and then I instantly type "Hi name" or "Hello name" and half way through the animation is done and the mail shows and it starts moving the cursor to the start so I get my initial text screwed up.

I hope you could give some kind of a fix a try :)

**Comment by houmark** [comment]
I know it's been a long time since this was discussed, but every single day I'm annoyed by the slowdown when I'm replying mails. I have this constant cursor dance and no matter how much I try to make myself wait for the animation to finish I fail and most times I will have to correct the text and cursor.

I would love if you would try some work on this and I would be happy to assist in testing.

**Comment by robertklep** [comment]
@houmark right now, I'm working on supporting High Sierra, which has priority.

If I have some spare time, I'll try and see if I can fix the issue that you're having.

**Comment by houmark** [comment]
I fully understand and appreciate that. Just wanted to note it again in case you come around this when refactoring for High Sierra 😏

**Comment by robertklep** [comment]
@houmark 

Try and see if this plugin solves (or at least mitigates) your issues:

https://github.com/AndreasVerhoeven/NoMailAppAnimationsOnMavericks/releases

It seems to work fine next to QF, and it should disable all animations, which are the main cause for the delayed cursor placement.

**Comment by houmark** [comment]
Unfortunately that only made animations jerky but there's still things going on with the window when it's doing modifications to the content of the mail. Thanks for the suggestion though...

**Comment by robertklep** [comment]
It should completely disable the animation, not make it jerky. For me, it works like that.

There's still some time between opening the window and setting the cursor, I'll see if I can fix that, but I think that's about as much fixing that I can do :(

**Comment by houmark** [comment]
If you could make the cursor stop moving completely I think there could be some progress.

To be honest, the only thing I use quotefix for is to make a custom attribution for the reply message so the time of the original mail is not shown. I work with clients in various time zones, so to not confuse I just remove the time.

<img width="406" alt="image 2017-08-27 at 8 48 11 pm" src="https://user-images.githubusercontent.com/467330/29756389-2db16678-8b69-11e7-8dab-c803db288087.png">

**Comment by robertklep** [comment]
@houmark can you try and see if 2.8.2 works better for you? https://github.com/robertklep/quotefixformac/releases/tag/v2.8.2

It removes _most_ of the cursor movements if quotefixing is disabled (custom attributions will still work):

<img width="718" alt="screen shot 2017-08-28 at 12 20 11" src="https://user-images.githubusercontent.com/1190916/29769580-50288b96-8beb-11e7-9ef9-4977b267f033.png">

**Comment by houmark** [comment]
Thanks for the attempt @robertklep. I re-installed the NoMailAppAnimations and also 2.8.2 of QuoteFix. Unfortunately while the cursor seems to be more still and the animations not jerky, the cursor still moves to the beginning of the mail body. Meanwhile my Mac takes a half a sec or so to open the mail but I'm typing in the meantime and when the body of the response appears quotefixed the cursor moves to the start and that's the point where my cursor in the middle of writing something like "Hello something" gets moved to the beginning so I either get a bad word and I have to delete and move the cursor around.

I know this is a special case but it's just a daily annoyance.

**Note:** I'm on a **brand new** latest generation MacBook Pro 15" (delivered and installed this friday night) maxed out specs with a clean install of macOS Sierra so the time I see opening and quotefixing the response should be as close to the minimal time anyone would see.

I attempted to do a small video / gif, but not too happy about sharing the content here and blurring out the sensible info was getting hard. I would have no problem passing the gif to you directly of you would like to visualize what's happening exactly.

**Comment by robertklep** [comment]
There are one or two cursor movements that are still required because of the attributions, and I can't get rid of them for now.

Feel free to e-mail me directly: robert DOT klep AT gmail DOT com

**Comment by houmark** [comment]
I understand that, and this may not be possible to solve unless one gives up other things, which may not even be possible either.

Will email you!
